Document the "retired" lifecycle stage (#14)
r-lib/lifecycle: vignettes/lifecycle.Rmd

<img src="figures/lifecycle.svg" width="75%" />

There are two development stages.

* <img src="figures/lifecycle-experimental.svg" alt = "Experimental" style="vertical-align:middle" /> This is a new feature that is in the very early stage of development. It is exported so users can start to use it and report feedback, but its interface and/or behaviour is likely to change in the future. It is generally best to avoid depending on experimental features.

* <img src="figures/lifecycle-maturing.svg" alt = "Maturing" style="vertical-align:middle" /> The interface and behaviour of a maturing feature has been roughed out, but finer details are likely to change. It still needs more feedback to find the optimal API.

> The probable end state of most functions is _stable_.

Reviewer: ```suggestion
The probable end state of most functions is _stable_:
```

* <img src="figures/lifecycle-stable.svg" alt = "Stable" style="vertical-align:middle" /> A feature is considered stable when the author is happy with its interface and behaviour. Major changes are unlikely, and breaking changes will occur gradually, through a deprecation process.

> Sometimes we are no longer certain that a feature follows or implements the right approach. In this case, we mark it as _questioning_. It can then either become _superseded_ once we have implemented an alternative, or _deprecated_. The difference between superseded and deprecated is that a superseded feature is maintained indefinitely for backward compatibility.

Reviewer: Not sure exactly where this should go, but need something like: Alternatively, we may realise that a questioning feature is actually important, and we move it back to stable.

Author: I think it's the right place. I added:

>  User feedback is usually important for taking these decisions and always welcome.

1. <img src="figures/lifecycle-questioning.svg" alt = "Questioning" style="vertical-align:middle" /> The author is no longer convinced that the feature is the optimal approach. However, there are no recommended alternatives yet.

> 1. <img src="figures/lifecycle-superseded.svg" alt = "Superseded" style="vertical-align:middle" /> A superseded feature is no longer under active development, and a known better alternative is available. However, it is indefinitely kept in the package for backward compatibility. The author only makes the necessary changes to ensure that the function continues working. No new features will be added, and only the most critical of bugs will be fixed.

Reviewer: ```suggestion
1. <img src="figures/lifecycle-superseded.svg" alt = "Superseded" style="vertical-align:middle" /> (Previously called retired) A superseded feature is no longer under active development, and a known better alternative is available. However, it is indefinitely kept in the package for backward compatibility. The author will only make the necessary changes to ensure that the function continues working. No new features will be added, and only critical bugs will be fixed.
```

1. <img src="figures/lifecycle-questioning.svg" alt = "Questioning" style="vertical-align:middle" /> The author is no longer convinced that the feature is the optimal approach. However, there are no recommended alternatives yet.

> When a feature is no longer under development, it can be either __retired__ or __deprecated__.

Reviewer: This doesn't feel quite right because I wouldn't say an stable function is under development either.

Do you want me to redraw the package lifecycle diagram for function components? I think having an overview picture would help.

Author: That'd be great!

Author: I think of stable functions as in development because users can request new features.

Reviewer: I think that will confuse people — I think we should be framing stable as the probable end state for most functions; it's only occasionally that functions will be retired or deprecated.

Reviewer 2: ```suggestion
Once the decision of discontinuing a feature has been made, it can be either __retired__ or __deprecated__.
```

Author: A retired feature is not discontinued, only the development is discontinued.

Reviewer 2: Oh, right. Can we say there's three development stages that sometimes lead to stability, which isn't primarily a development stage, and that stable functions are occasionally retired?

Author: I like the idea of making it clearer that retired functions are not discontinued. Maybe the stage should be `stable-retired`?

This comes up often, I had to mention on SO just 2 days ago that plyr wasn't deprecated.

Reviewer 2: I'm suggesting to mention "stable" and "retired" in the same section in this document, which (I think) is similar in spirit to what you suggest? This would make "retired" a form of "stable", but I'm not really in love with the combo word.

What about ![immortal](https://img.shields.io/badge/lifecycle-immortal-gold.svg)?
